### PR TITLE
[3006.x] Preserve socket between requests

### LINKED ISSUE #64627: [BUG] 3005.1 minion does not close connections to port 4506 (most likely affects 3006 too)
**Description**
3005.1 minion does not close connections to port 4506 (based on the current code most likely it affects 3006 too). As mentioned at the bottom of https://github.com/saltstack/salt/issues/64552 because of the dubious changes made to the minion code recently the minion keeps the connection to master port 4506 open after it sends a message. That connection is only closed right before it reopens it while sending another message which is a total misuse (and obviously that reopened connection is then left open until the cycle repeats). I elaborated more on it in https://github.com/saltstack/salt/issues/64552 (last few comments there).
To depict it with a metaphor, it works as if you always kept your front door wide open and only closed it when you go in or out just to leave it open again.

**Setup**
Any 3005.1 or 3006 setup should do to reproduce this behaviour.

**Steps to Reproduce the behavior**
Run "netstat -an | grep :4506" on a minion, run a job on it from the master and run netstat again. Observe a new connection kept in ESTABLISHED state and the previous one in TIME_WAIT. Use tcpdump and or iptables (with -j LOG) to watch the traffic, although in https://github.com/saltstack/salt/issues/64552 I explained at length why and how this happens.

**Expected behavior**
Connection to master port 4506 is closed after a message is sent.

### PR DESCRIPTION
**Pull request by dwoz**
### What does this PR do?

Fixes #64627

Also includes fix for tests that broke when #65053 was merged.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### CODE REVIEW

File: salt/transport/zeromq.py [hunk 1]
```diff
@@ -515,12 +516,14 @@ def __init__(self, opts, addr, linger=0, io_loop=None):
         self.context = zmq.Context()
 
         self.send_queue = []
-        # mapping of message -> future
-        self.send_future_map = {}
 
         self._closing = False
+        self.future = None
```
Comment (s0undt3ch) [inline salt/transport/zeromq.py:521]:
```suggestion
        self._future = None
```
Right?

File: salt/transport/zeromq.py [hunk 2]
```diff
@@ -577,19 +577,20 @@ def _init_socket(self):
...
 
         :raises: SaltReqTimeoutError
         """
-        future = self.send_future_map.pop(message, None)
-        # In a race condition the message might have been sent by the time
-        # we're timing it out. Make sure the future is not None
-        if future is not None:
-            future.set_exception(SaltReqTimeoutError("Message timed out"))
+        try:
+            self._future = None
```
Comment (s0undt3ch) [inline salt/transport/zeromq.py:590]: Why would setting an attribute to `None` raise a `ValueError`?

File: tests/pytests/functional/transport/zeromq/test_request_client.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    minion_opts["master_uri"] = f"tcp://127.0.0.1:{port}"
+
+    ctx = zmq.Context()
+    socket = ctx.socket(zmq.REP)
+    socket.bind(minion_opts["master_uri"])
+    stream = zmq.eventloop.zmqstream.ZMQStream(socket, io_loop=io_loop)
+
+    @salt.ext.tornado.gen.coroutine
+    def req_handler(stream, msg):
+        print(f"REQ HANDLER {stream.send} {msg}")
```
Comment (s0undt3ch) [inline tests/pytests/functional/transport/zeromq/test_request_client.py:28]:
```suggestion
        print(f"REQ HANDLER {stream.send} {msg}")
```
Log this or remove print.

File: tests/pytests/functional/transport/zeromq/test_request_client.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        yield stream.send(msg[0])
+
+    stream.on_recv_stream(req_handler)
+
+    request_client = salt.transport.zeromq.RequestClient(minion_opts, io_loop)
+
+    rep = await request_client.send(b"foo")
+    req_socket = request_client.message_client.stream.socket
+    rep = await request_client.send(b"foo")
+    assert req_socket == request_client.message_client.stream.socket
```
Comment (s0undt3ch) [inline tests/pytests/functional/transport/zeromq/test_request_client.py:38]:
```suggestion
    assert req_socket is request_client.message_client.stream.socket
```
We want to make the the socket reference is the same as the the current reference, not equal to, right?

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

File: tests/pytests/functional/transport/zeromq/test_request_client.py (hunk 3 above)
Comment (dwoz) [reply tests/pytests/functional/transport/zeromq/test_request_client.py]: fixed

File: tests/pytests/functional/transport/zeromq/test_request_client.py (hunk 4 above)
Comment (dwoz) [reply tests/pytests/functional/transport/zeromq/test_request_client.py]: fixed

File: salt/transport/zeromq.py (hunk 2 above)
Comment (dwoz) [reply salt/transport/zeromq.py]: fixed

File: salt/transport/zeromq.py (hunk 1 above)
Comment (dwoz) [reply salt/transport/zeromq.py]: fixed

**Review by s0undt3ch — APPROVED** [review]

**Review by dmurphy18 — APPROVED** [review]

**Review by twangboy — APPROVED** [review]

**Pull request merged**